include/rmm/mr/device/polymorphic_allocator.hpp

```diff
@@ -0,0 +1,278 @@
+};
+
+template <typename T, typename U>
+bool operator==(polymorphic_allocator<T> const& lhs, polymorphic_allocator<U> const& rhs)
+{
+  return lhs.resource()->is_equal(*rhs.resource());
+}
+
+template <typename T, typename U>
+bool operator!=(polymorphic_allocator<T> const& lhs, polymorphic_allocator<U> const& rhs)
+{
+  return not(lhs == rhs);
+}
+
+/**
+ * @brief Adapts a stream ordered allocator to provide a standard `Allocator` interface
+ *
+ * A stream-ordered allocator (i.e., `allocate/deallocate` use a `cuda_stream_view`) cannot be used
+ *in an interface that expects a standard C++ `Allocator` interface. `stream_allocator_adaptor`
+ *wraps a stream-ordered allocator and a stream to provide a standard `Allocator` interface. The
+ *adaptor uses the wrapped stream in calls the underlying allocator's `allocate` and `deallocate`
+ * functions.
+ *
+ * Example:
+ *\code{c++}
+ * my_stream_ordered_allocator<int> a{...};
+ * cuda_stream_view s = // create stream;
+ *
+ * auto adapted = make_stream_allocator_adaptor(a, s);
+ *
+ * // Allocates storage for `n` int's on stream `s`
+ * int * p = std::allocator_traits<decltype(adapted)>::allocate(adapted, n);
+ *\endcode
+ *
+ * @tparam Allocator Stream ordered allocator type to adapt
+ */
+template <typename Allocator>
+class stream_allocator_adaptor {
+ public:
+  using value_type = typename std::allocator_traits<Allocator>::value_type;
+
+  stream_allocator_adaptor() = delete;
+
+  /**
+   * @brief Construct a `stream_allocator_adaptor` using `a` as the underlying allocator.
+   *
+   * @note: The `stream` must not be destroyed before the `stream_allocator_adaptor`, otherwise
+   * behavior is undefined.
+   *
+   * @param a The stream ordered allocator to use as the underlying allocator
+   * @param stream The stream used with the underlying allocator
+   */
+  stream_allocator_adaptor(Allocator const& a, cuda_stream_view stream) : alloc_{a}, stream_{stream}
+  {
+  }
+
+  /**
+   * @brief Construct a `stream_allocator_adaptor` using `other.underlying_allocator()` and
+   * `other.stream()` as the underlying allocator and stream.
+   *
+   * @param other The other `stream_allocator_adaptor` whose underlying allocator and stream will be
+   * copied
+   */
+  stream_allocator_adaptor(stream_allocator_adaptor<Allocator> const& other) = default;
+
+  /**
+   * @brief Construct a `stream_allocator_adaptor` using `other.underlying_allocator()` and
+   * `other.stream()` as the underlying allocator and stream.
+   *
+   * @tparam OtherAllocator Type of `other`'s underlying allocator
+   * @param other The other `stream_allocator_adaptor` whose underlying allocator and stream will be
+   * copied
+   */
+  template <typename OtherAllocator>
+  stream_allocator_adaptor(stream_allocator_adaptor<OtherAllocator> const& other)
+    : stream_allocator_adaptor{other.underlying_allocator(), other.stream()}
+  {
+  }
+
+  /**
+   * @brief Rebinds the allocator to the specified type.
```

Comment:
Why does `stream_allocator_adaptor` have rebind and `polymorphic_allocator` doesn't?

Reply:
The default `rebind` provided by `std::allocator_traits` works for `polymorhpic_allocator` but not `stream_adaptor_allocator` because the former templates on `T` and the latter templates on `Allocator`.